Adding CloudFoundryPaasLocation and its resolver.

kiuby88 opened the pull request:
A simple `CloudFoundryPaasLocation` and its resolver were added.

It it s very simple PR, but it will allows to  `PaasApplication` management to be reached in the next step.

kiuby88 committed:
Adding CloudFoundryPaasLocation and its resolver.

Pom file was optimized.

kiuby88 commented:
@andreaturli take a look, please. I think it is the first step to reach the new design which we are developing ;)

andreaturli commented on pom.xml:
```diff
@@ -25,20 +23,17 @@
     <parent>
         <groupId>org.apache.brooklyn</groupId>
         <artifactId>brooklyn-downstream-parent</artifactId>
-        <version>0.10.0-SNAPSHOT</version> <!-- BROOKLYN_VERSION -->
```
leave the comment

kiuby88 replied:
Oops

andreaturli commented on pom.xml:
```diff
@@ -48,42 +43,38 @@
+        <dependency>
+            <groupId>org.apache.brooklyn</groupId>
+            <artifactId>brooklyn-all</artifactId>
+            <version>${brooklyn.version}</version>
```
use

```
<version>${project.parent.version}</version>
```

instead of

```
<version>${brooklyn.version}</version>
```

kiuby88 replied:
I did not know about that. Thanks.

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,42 @@
+    public static ConfigKey<String> CF_ORG = ConfigKeys.newStringConfigKey("org");
+    public static ConfigKey<String> CF_ENDPOINT = ConfigKeys.newStringConfigKey("endpoint");
+    public static ConfigKey<String> CF_SPACE = ConfigKeys.newStringConfigKey("space");
+
```
I think it is better to move those properties to a `PaasLocationConfig` interface.
It may even `extends CloudLocationConfig` but not sure yet.

kiuby88 replied:
I do not agree, because this properties are closed to CloudFoundry but I would agree about moving them to `CloudFoundryPaasLocationConfig`.

andreaturli replied:
try to generalize them and put the generic configKey in `PaasLocationConfig` and some others into `CloudFoundryPaasLocationConfig` 
Thinking at it better you should reuse as much as possible the properties defined in `LocationConfigKeys`

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+ * It contains the parameters needed to configure an CloudFoundryPaasLocation. For example, the ConfigKey
+ * {@link #REQUIRED_MEMORY} allows to specify a initial memory amount used in a location.
+ */
+public interface CloudFoundryPaasLocationConfig {
```
I think this should `extends PaasLocationConfig`

kiuby88 replied:
I can create that interface which would contain somethings like:

```
@SetFromFlag("paas.memory")
ConfigKey<Integer> REQUIRED_MEMORY = ConfigKeys.newIntegerConfigKey(
            "paas.profile.memory", "Memory allocated for the application (MB)", 512);
```

Then, `CloudFoundryPaasLocationConfig` will implement that interface and should contain closed ConfigKeys that will be closed to CloudFoundry such as `CF_USER` or `CF_SPACE`

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public String getPaasProviderName() {
+        return "CloudFoundry";
```
not sure about this static value;
also I think you should be consistent with the rest of the code `cloudfoundry` is used elsewhere

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+import org.apache.brooklyn.util.core.flags.SetFromFlag;
+
+/**
+ * It contains the parameters needed to configure an CloudFoundryPaasLocation. For example, the ConfigKey
```
move `For example ...` above `REQUIRED_MEMORY`

add javadoc for all the config keys or remove from all of them

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+    ConfigKey<Integer> REQUIRED_INSTANCES = ConfigKeys.newIntegerConfigKey(
+            "cloudfoundry.profile.instances", "Number of instances of the application", 1);
+
+    @SetFromFlag("cloudfoundry.instances")
```
"cloudfoundry.instances"?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+    ConfigKey<Integer> REQUIRED_MEMORY = ConfigKeys.newIntegerConfigKey(
+            "cloudfoundry.profile.memory", "Memory allocated for the application (MB)", 512);
+
+    @SetFromFlag("cloudfoundry.instances")
```
"cloudfoundry.instances" ?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolver.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public LocationSpec<? extends Location> newLocationSpecFromString(String s, Map<?, ?> map, LocationRegistry locationRegistry) {
+        return LocationSpec
+                .create(CloudFoundryPaasLocation.class)
```
weird indentation, just one line isn't better?

andreaturli commented on src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolverTest.java:
```diff
@@ -0,0 +1,93 @@
+    private LocalManagementContext managementContext;
+    private BrooklynProperties brooklynProperties;
+
+    private final String LOCATION_SPEC_NAME = "cloudfoundry-instance";
```
what is that? maybe `my-cloudfoundry`. if so rename `LOCATION_SPEC_NAME` to `MY_PAAS_LOCATION`

kiuby88 commented:
Thanks for the review, @andreaturli. Let me address them

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolver.java:
```diff
@@ -0,0 +1,71 @@
+        return LocationSpec
+                .create(CloudFoundryPaasLocation.class)
+                .configure(map);
+    }
```
I think it could be interesting to allow also LocationSpecString like:

```
cloudfoundry[:endpoint:org:space]
cloudfoundry[:provider:org:space]
```

the properties in square brackets are optional.

The second provider can be `provider: pivotal` so in that case the endpoint is inferred from `pivotal` provider but let's keep that in mind for a subsequent PR, in case.

kiuby88 replied:
Great idea.

andreaturli commented on src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolverTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+
+}
```
add an empty line

andreaturli commented on src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolverTest.java:
```diff
@@ -0,0 +1,93 @@
+        assertEquals(cloudFoundryPaasLocation.getConfig(CloudFoundryPaasLocation.CF_ORG), ORG);
+        assertEquals(cloudFoundryPaasLocation.getConfig(CloudFoundryPaasLocation.CF_SPACE), SPACE);
+
+        assertEquals(cloudFoundryPaasLocation.getConfig(CloudFoundryPaasLocationConfig.REQUIRED_DISK),
```
not sure these assertions make sense as the default value is not defined.

andreaturli commented on src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolverTest.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Test
+    public void testCloudFoundryTakesProvidersScopedProperties() {
+        CloudFoundryPaasLocation cloudFoundryPaasLocation = resolve(LOCATION_SPEC_NAME);
```
I think it is better to extract all config

```
Map<String, Object> conf = resolve(LOCATION_SPEC_NAME).config().getBag().getAllConfig();
```

and make assertions on them

kiuby88 committed:
Generalization of CloudFoundryPasLocation's configkey.
Minor fix in:
- pom file
- the provider name of CloudFoundryPaasLocation
- style indentation

kiuby88 commented:
@andreaturli thanks for your useful review. I have addressed all your suggestions.

kiuby88 commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/paas/PaasLocationConfig.java:
```diff
@@ -0,0 +1,52 @@
+    @SetFromFlag("endpoint")
+    public static ConfigKey<String> CLOUD_ENDPOINT = LocationConfigKeys.CLOUD_ENDPOINT;
+
+    @SetFromFlag("paas.instances")
```
@andreaturli, do you think it would be better something like
`@SetFromFlag("profile.instances")` ?

kiuby88 commented:
@andreaturli, Ready!! (I think)
